### PR TITLE
MGMT-9809: Modify existing python code in the swarm tool to use the combined agent

### PR DESCRIPTION
**Pull request by ori-amizur**
Spawn combined-agent once for swarm, and create an agent by api call
instead of spawning an agent in a container.
When waiting for agent to terminate, poll its status from the
combined-agent, instead of waiting the process to exit.
When the swarm terminates, exit the combined-agent, by calling exit api
call.

/cc @omertuc

### CODE REVIEW

File: combined_agent.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import subprocess
+import waiting
+
+DEFAULT_AGENT_IMAGE = "quay.io/oamizur/assisted-swarm:latest"
+DEFAULT_STDOUT_FILE = "/tmp/combined-agent.stdout"
+DEFAULT_STDERR_FILE = "/tmp/combined-agent.stderr"
+
+class CombinedAgent(object):
```
Comment (omertuc) [inline combined_agent.py:8]:
```suggestion
class CombinedAgent:
```

`object` inheritance not needed in Python3

File: combined_agent.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    def __init__(self, executor, swarm_client):
```
Comment (omertuc) [inline combined_agent.py:9]: Could you add type hints here for those two?

File: hack/Dockerfile.python-client [hunk 3]
```diff
@@ -0,0 +1,5 @@
+FROM quay.io/edge-infrastructure/swagger-codegen-cli:2.4.18
```
Comment (omertuc) [inline hack/Dockerfile.python-client:1]: Could you rename this file to `Dockerfile.python-client-generate`? Current name is a bit misleading

File: combined_agent.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            "podman run --rm --net=host --privileged -v /var/log:/var/log -v /run/udev:/run/udev -v /dev/disk:/dev/disk "
+            "-v /run/systemd/journal/socket:/run/systemd/journal/socket -v /var/log:/host/var/log:ro -v /proc/meminfo:/host/proc/meminfo:ro "
+            "-v /sys/kernel/mm/hugepages:/host/sys/kernel/mm/hugepages:ro -v /proc/cpuinfo:/host/proc/cpuinfo:ro -v /etc/mtab:/host/etc/mtab:ro "
+            "-v /sys/block:/host/sys/block:ro -v /sys/devices:/host/sys/devices:ro -v /sys/bus:/host/sys/bus:ro -v /sys/class:/host/sys/class:ro "
+            "-v /run/udev:/host/run/udev:ro -v /dev/disk:/host/dev/disk:ro -v /opt:/opt:rw " + agent_image
+        )
+        with open(stdout_fname, "ab") as stdout_file:
+            with open(stderr_fname, "ab") as stderr_file:
+                self.executor.Popen(
+                        command.split(),
```
Comment (omertuc) [inline combined_agent.py:24]:
```suggestion
    def spawn(
        self,
        agent_image=DEFAULT_AGENT_IMAGE,
        stdout_fname=DEFAULT_STDOUT_FILE,
        stderr_fname=DEFAULT_STDERR_FILE,
    ):
        command = (
            "podman",
            "run",
            "--rm",
            "--net=host",
            "--privileged",
            "--volume=/var/log:/var/log",
            "--volume=/run/udev:/run/udev",
            "--volume=/dev/disk:/dev/disk",
            "--volume=/run/systemd/journal/socket:/run/systemd/journal/socket",
            "--volume=/var/log:/host/var/log:ro",
            "--volume=/proc/meminfo:/host/proc/meminfo:ro",
            "--volume=/sys/kernel/mm/hugepages:/host/sys/kernel/mm/hugepages:ro",
            "--volume=/proc/cpuinfo:/host/proc/cpuinfo:ro",
            "--volume=/etc/mtab:/host/etc/mtab:ro",
            "--volume=/sys/block:/host/sys/block:ro",
            "--volume=/sys/devices:/host/sys/devices:ro",
            "--volume=/sys/bus:/host/sys/bus:ro",
            "--volume=/sys/class:/host/sys/class:ro",
            "--volume=/run/udev:/host/run/udev:ro",
            "--volume=/dev/disk:/host/dev/disk:ro",
            "--volume=/opt:/opt:rw",
            agent_image,
        )
        with open(stdout_fname, "ab") as stdout_file:
            with open(stderr_fname, "ab") as stderr_file:
                self.executor.Popen(
                    command,
                    stdin=subprocess.DEVNULL,
                    stdout=stdout_file,
                    stderr=stderr_file,
                )
        self.wait_for_health()

```

File: combined_agent.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                        stderr=stderr_file
+                    )
+        self.wait_for_health()
+
+    def wait_for_health(self):
+        def get_health_status():
+            try:
+                self.swarm_client.health()
+                return True
+            except:
```
Comment (omertuc) [inline combined_agent.py:36]:
```suggestion
            except Exception:
```

To avoid capturing `KeyboardInterrupt`

File: combined_agent.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+                return False
+        waiting.wait(get_health_status,
+                     sleep_seconds=1,
+                     timeout_seconds=10)
```
Comment (omertuc) [inline combined_agent.py:40]: May need much more time than this when doing a pull from a registry

File: swarm_api.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python3
+
+from __future__ import print_function
+import assisted_swarm_client.assisted_swarm as assisted_swarm
+
+
+def new_swarm_client():
+    configuration = assisted_swarm.Configuration()
+    configuration.host = 'http://127.0.0.1:5566/api/assisted-swarm'
```
Comment (omertuc) [inline swarm_api.py:9]:
```suggestion
    configuration.host = 'http://localhost:5566/api/assisted-swarm'
```

nit

File: Makefile [hunk 8]
```diff
@@ -1,6 +1,6 @@
 ASSISTED_SWARM = build/assisted-swarm
 CONTAINER_COMMAND := $(or $(CONTAINER_COMMAND),docker)
-IMAGE := $(or $(IMAGE),quay.io/openshift/assisted-swarm:latest)
+IMAGE := $(or $(IMAGE),quay.io/oamizur/assisted-swarm:latest)
```
Comment (omertuc) [inline Makefile:3]: Did you commit this intentionally?

File: agent.py [hunk 9]
```diff
@@ -75,7 +78,9 @@ def __init__(
         self,
         swarm_agent_config: SwarmAgentConfig,
         cluster_agent_config: ClusterAgentConfig,
+        swarm_client: SwarmApi,
```
Comment (omertuc) [inline agent.py:81]: Pass this as part of `swarm_agent_config: SwarmAgentConfig`

Comment (ori-amizur) [reply agent.py:81]: It is not a configuration

File: cluster.py [hunk 10]
```diff
@@ -46,7 +47,8 @@ class ClusterConfig:
 
 
 class Cluster(RetryingStateMachine, WithContainerConfigs):
-    def __init__(self, cluster_config: ClusterConfig, swarm_agent_config: SwarmAgentConfig):
+    def __init__(self, cluster_config: ClusterConfig, swarm_agent_config: SwarmAgentConfig, swarm_client: SwarmApi):
+        self.swarm_client = swarm_client
```
Comment (omertuc) [inline cluster.py:51]: Pass this as part of `cluster_config: ClusterConfig`

Comment (ori-amizur) [reply cluster.py:51]: It is not a configuration

File: Makefile (hunk 8 above)
Comment (ori-amizur) [reply Makefile:3]: For now, until we will have an image available in openshift

File: agent.py (hunk 9 above)
Comment (omertuc) [reply agent.py:81]: `ClusterAgentConfig` is meant for cluster-specific parameters that the agent needs, `SwarmAgentConfig` is meant for generic utilities / parameters the agent needs (see it has kube_cache, executor, etc). The client object fits perfectly in there

File: cluster.py (hunk 10 above)
Comment (omertuc) [reply cluster.py:51]: Actually if you do the other comment, the cluster won't even have to be aware of this object - the swarm creates the SwarmAgentConfig and it's passed as an opaque object to all agents within the cluster

File: agent.py (hunk 9 above)
Comment (omertuc) [reply agent.py:81]: See also https://github.com/openshift-assisted/assisted-swarm/pull/2/files#r862448691

File: Makefile (hunk 8 above)
Comment (omertuc) [reply Makefile:3]: Alright. Should probably be edge-infrastructure though

File: skipper.yaml [hunk 11]
```diff
@@ -4,6 +4,8 @@ build-container-image: assisted-swarm-build
 volumes:
   # cache
   - $HOME/.cache/go-build:/go/pkg/mod
+  # push
```
Comment (omertuc) [inline skipper.yaml:7]:
```suggestion
  # Registry credentials
```

File: agent.py [hunk 12]
```diff
@@ -239,6 +244,18 @@ def provisioned_bmh(self, next_state):
 
         return self.state
 
+    def wait_for_completion(self, id):
+        def has_agent_completed():
+            response = self.swarm_client.get_agent(id)
+            return response.status == "Terminated"
```
Comment (omertuc) [inline agent.py:250]:
```suggestion
            return response.status == AgentStatus.TERMINATED
```

File: agent.py [hunk 13]
```diff
@@ -16,6 +17,8 @@
 from swarmexecutor import SwarmExecutor
 from swarmkubecache import SwarmKubeCache
 from withcontainerconfigs import WithContainerConfigs
+import assisted_swarm_client.assisted_swarm as assisted_swarm
+from assisted_swarm_client.assisted_swarm import SwarmApi
```
Comment (omertuc) [inline agent.py:21]:
```suggestion
from assisted_swarm_client.assisted_swarm import SwarmApi, AgentStatus
```

File: agent.py (hunk 9 above)
Comment (ori-amizur) [reply agent.py:81]: Can do that, but I still think that in this case there is a difference between configuration class and functional class

File: combined_agent.py (hunk 1 above)
Comment (ori-amizur) [reply combined_agent.py:8]: OK

File: combined_agent.py (hunk 4 above)
Comment (ori-amizur) [reply combined_agent.py:24]: OK

File: agent.py (hunk 9 above)
Comment (omertuc) [reply agent.py]: I agree that Config is not the best name, but that class was meant to give the agent all the dependencies it needs to function, including functional things (like KubeCache and the executor).

**Comment by omertuc** [comment]
So, is this working well for a lot of agents?

**Pull request merged**